coramin/relaxations/mccormick.py

```diff
@@ -158,3 +86,193 @@ def _build_mccormick_relaxation(b, x, y, w, relaxation_side=RelaxationSide.BOTH,
                 # see the doc string for this method - only adding one over and one under-estimator
                 if xlb != -math.inf and yub != math.inf:
                     b.relaxation.add(w <= x*yub + xlb*y - xlb*yub)
+
+
+def _build_pw_mccormick_relaxation(b, x, y, w, x_pts, relaxation_side=RelaxationSide.BOTH):
+    """
+    This function creates piecewise envelopes to relax "w = x*y". Note that the partitioning is done on "x" only.
+    This is the "nf4r" from Gounaris, Misener, and Floudas (2009).
+
+    Parameters
+    ----------
+    b: pyo.ConcreteModel or pyo.Block
+    x: pyomo.core.base.var._GeneralVarData
+        The "x" variable in x*y
+    y: pyomo.core.base.var._GeneralVarData
+        The "y" variable in x*y
+    w: pyomo.core.base.var._GeneralVarData
+        The "w" variable that is replacing x*y
+    x_pts: list of floats
+        A list of floating point numbers to define the points over which the piecewise representation will generated.
+        This list must be ordered, and it is expected that the first point (x_pts[0]) is equal to x.lb and the
+        last point (x_pts[-1]) is equal to x.ub
+    relaxation_side : minlp.minlp_defn.RelaxationSide
+        Provide the desired side for the relaxation (OVER, UNDER, or BOTH)
+
+    """
+    xlb = x_pts[0]
+    xub = x_pts[-1]
+    ylb, yub = tuple(_get_bnds_list(y))
+
+    check_var_pts(x, x_pts=x_pts)
+    check_var_pts(y)
+
+    if x.is_fixed() and y.is_fixed():
+        b.xy_fixed_eq = pyo.Constraint(expr= w == pyo.value(x) * pyo.value(y))
+    elif x.is_fixed():
+        b.x_fixed_eq = pyo.Constraint(expr= w == pyo.value(x) * y)
+    elif y.is_fixed():
+        b.y_fixed_eq = pyo.Constraint(expr= w == x * pyo.value(y))
+    elif len(x_pts) == 2:
+        _build_mccormick_relaxation(b, x=x, y=y, w=w, relaxation_side=relaxation_side)
+    elif xlb == -math.inf or xub == math.inf or ylb == -math.inf or yub == math.inf:
+        _build_mccormick_relaxation(b, x=x, y=y, w=w, relaxation_side=relaxation_side)
+    else:
+        # create the lambda variables (binaries for the pw representation)
+        b.interval_set = pyo.Set(initialize=range(1, len(x_pts)))
+        b.lam = pyo.Var(b.interval_set, within=pyo.Binary)
```

Comment:
Is the abbreviation necessary?  Why not `b.lambda`?

Reply:
No, the abbreviation is not necessary. Good point.

Reply:
Oh, `lambda` is a builtin function name. Although the namespace makes this work, I prefer not to use builtin function names for anything...

Reply:
Hmm... I would actually be OK with `lambda` (we have a similar issue with the `set` module in Pyomo).  Alternatively, you could use `_lambda` or `lambda_`.